README.md
BusyBird: a multi-level Web-based timeline viewer
=================================================

[![Build Status](https://travis-ci.org/debug-ito/busybird.svg?branch=master)](https://travis-ci.org/debug-ito/busybird)

BusyBird is a personal Web-based timeline viewer application.
You can think of it as a Twitter client, but BusyBird is more generic and focused on viewing.

BusyBird accepts data called **Statuses** from its RESTful Web API.
The received statuses are stored to one or more **Timelines** .
You can view those statuses in a timeline by a Web browser.

For more information, visit https://metacpan.org/pod/BusyBird

SCREENSHOTS
-----------

https://github.com/debug-ito/busybird/wiki/Screenshots

QUICK START
-----------

Example in Ubuntu Linux.

- Install `make` and `curl`

        $ sudo apt-get install build-essential curl

- Install

        $ curl -L http://cpanmin.us/ | perl - -n BusyBird
        $ export PERL5LIB="$HOME/perl5/lib/perl5:$PERL5LIB"
        $ export PATH="$HOME/perl5/bin:$PATH"

- Run

        $ busybird
        Twiggy: Accepting connections at http://127.0.0.1:5000/

- Open timelines

        $ firefox http://localhost:5000/

- Post a status

        $ curl -d '{"text":"hello, world!"}' http://localhost:5000/timelines/home/statuses.json

See https://metacpan.org/pod/BusyBird for detail.


TUTORIAL
--------

See https://metacpan.org/pod/BusyBird::Manual::Tutorial


TRY WITHOUT INSTALLATION
------------------------

You can try BusyBird without installing it. This is recommended if you
try a development version. (You need `cpanm` command. See
[App::cpanminus](https://metacpan.org/pod/App::cpanminus) for detail).

    $ git clone https://github.com/debug-ito/busybird.git
    $ cd busybird
    $ cpanm Module::Build::Prereqs::FromCPANfile
    $ cpanm --installdeps .
    $ perl Build.PL
    $ ./Build
    $ ./Build test

...and to start BusyBird, type

    $ perl -Iblib/lib blib/script/busybird


AUTHOR
------

Toshio Ito

* https://github.com/debug-ito
* debug.ito [at] gmail.com


Changes
Revision history for Perl module BusyBird

0.12    2014-12-??
        - When doing "use Exporter qw(import)" you should require Exporter 5.57,
          since that was the first version that could export the import() function.
        - Added "use strict" and "use warnings" to the modules that didn't have them.

0.11    2014-10-22
        [ENHANCEMENTS]
        - Now it renders "busybird.user_permalink" field as the link URL for the user.
          This is an experimental feature for now.

0.10    2014-10-15
        [ENHANCEMENTS]
        - Web API: add "only_statuses" query parameter to GET /timelines/{timeline}/statues.json
        - Add per-timeline config params: "acked_statuses_load_count" and "default_level_threshold".
        - Now it warns you when you give unknown config parameters.
        [DOCUMENTATION]
        - BusyBird::SafeData is now public class. End-users are free to use it.
        [PACKAGING]
        - Drop dependency on Test::Exception. Use Test::Fatal instead.
        - Add dependency on EV. This is necessary to avoid a bug in autovivification
          ( https://rt.cpan.org/Public/Bug/Display.html?id=99458 )

0.09    2014-09-28
        [BUG FIX]
        - Now it is more tolerant of statuses with unusual structure.
        - Now Text::Xslate's warnings are reported through BusyBird::Log.
        [API CHANGE]
        - The default attached_image_urls_builder now ignores Twitter Entities
          whose 'type' field is set to NOT 'photo'. This change makes it robust
          to future changes in Twitter API. I think this change is backword-compatible in most cases.
        - Refactoring and API revision of BusyBird::Main::PSGI::View.
        [PACKAGING]
        - Add 'Data::Diver' to dependency.

0.08    2014-09-15
        [ENHANCEMENT]
        - Now it renders image URLs attached to statuses.
          To configure how the attached images should be rendered,
          set "attached_image_*" config parameters.
        [UI CHANGE]
        - Revise HTML mark-ups and CSS. Visually it makes little difference.
        [DOCUMENTATION]
        - Make it a little more friendly to non-Perl users.

0.07    2014-08-17
        [PACKAGING]
        - Migrate to Module::Build::Prereqs::FromCPANfile.

0.06    2014-08-14
        [API CHANGE]
        - Timeline and StatusStorage: now contains() method accepts status objects with no "id" field.
          They are always treated as "not contained". Previously the contains() method croaked when
          ID-less statuses were passed. This behavior has been changed.
        [BUG FIX]
        - Now BusyBird::Filter exports functions only by request. Previously they were automatically exported.
        [ENHANCEMENT]
        - Add "hidden" per-timeline config parameter.
        - Add filter_grep() function to BusyBird::Filter.
        - Add make_tracking() function to BusyBird::Util.

0.05    2014-08-13
        [BUG FIX]
        - Fix test t/WebAPI.t. It failed when consecutive acks occurred at different timestamps.
        [ENHANCEMENT]
        - StatusStorage::SQLite enhancements.
        - Add get_timeline_names() method.
        - Re-implement contains() method. This boosts its performance at about 350%.
        - Use SQL::Maker's strict mode with SQL::QueryMaker. This prevents unexpected query
          injection in form of complex data structure (often as JSON strings).
          This also improves performance a little.
        [DOCUMENTATION]
        - Fix Util.pm: sort_statuses() signature was wrong.

0.04    2014-07-28
        [BUG FIX]
        - Timelines whose name contains "/" could be created and installed, but were not accessible
          via HTTP. Now those timelines can be created, but cannot be installed in Main object.
          (gh #3)
        [ENHANCEMENT]
        - Add create_timeline() method to BusyBird::Main.
        [PACKAGING]
        - BusyBird::DateTime::Format now has a separate distribution. This makes it easier
          to develop and release input/filter modules that use the Format module.

0.03    2014-07-24
        [UI CHANGE]
        - Use Web Notifications to notify new statuses.
          To disable this feature, set "timeline_web_notifications" config parameter to "off"
          (or anything other than "simple"). (gh #2 thanks to Keisuke Minami)
        - Add favicon.
        - Remove horizontal paddings around statuses in small screens. (gh #1)
        [DOCUMENTATION]
        - Minor fixes.

0.02    2014-06-18
        [BUG FIX]
        - Fix test (t/StatusStorage_SQLite.t) so BSD systems can pass it.
        - Drop dependency on CPS::Functional from cpanfile.
        [UI CHANGE]
        - All controls in timeline navbar should now fit in 320px width screen.
        - Now it doesn't set the status permalink to Twitter if the status's ID is not an integer.
        [DOCUMENTATION]
        - Add "QUICK START" section.
        - Install BusyBird under user's home directory.

0.01    2014-06-15
        - First version, released on an unsuspecting world.


Makefile.PL
use inc::Module::Install;

name     'BusyBird';
all_from 'lib/BusyBird.pm';
author   q{Toshio ITO <toshioito@cpan.org>};
license  'perl';

requires 'Async::Queue', '0.02';
requires 'Async::Selector', '1.03';
requires 'Async::Selector::Aggregator', '1.02';
requires 'CPS', '0.15';
requires 'CPS::Functional', '0.15';
requires 'Carp', '0';
requires 'DateTime', '0.70';
requires 'DateTime::Duration', '0';
requires 'DateTime::Format::Strptime', '0';
requires 'DateTime::TimeZone', '1.42';
requires 'Encode', '0';
requires 'Exporter', '0';
requires 'File::ShareDir', '0';
requires 'File::Spec', '0';
requires 'JSON', '0';
requires 'List::Util', '1.24';
requires 'Plack::Builder', '0';
requires 'Plack::Request', '0';
requires 'Router::Simple', '0';
requires 'Scalar::Util', '1.24';
requires 'Storable', '2.27';
requires 'Test::Builder', '0';
requires 'Test::Exception', '0';
requires 'Test::More', '0.98';
requires 'Text::Xslate', '1.60';
requires 'Tie::IxHash', '0';
requires 'Time::HiRes', '1.9720';
requires 'Try::Tiny', '0.10';
requires 'autovivification', '0.10';
requires 'JavaScript::Value::Escape', '0';
requires 'Cache::Memory::Simple', '1.01';
requires 'URI::Escape', '0';
requires 'perl', '5.10.0';
requires 'sort', '0';

test_requires 'FindBin', '0';
test_requires 'HTTP::Request', '6.00';
test_requires 'Plack::Test', '0';
test_requires 'Test::Builder', '0';
test_requires 'Test::Exception', '0.31';
test_requires 'Test::Memory::Cycle', '1.04';
test_requires 'Test::MockObject', '1.09';
test_requires 'Test::More', '0.98';
test_requires 'Test::Warn', '0.24';
test_requires 'HTML::TreeBuilder', '5';
test_requires 'HTML::TreeBuilder::XPath', '0.12';
test_requires 'JSON', '0';


## auto_install;

repository 'https://github.com/debug-ito/busybird';

tests_recursive;
install_share;

WriteAll;
